src/qutip_qip/device/cavityqed.py

```diff
@@ -248,7 +248,14 @@ def eliminate_auxillary_modes(self, U):
         psi_proj = tensor(
             [basis(self.num_levels, 0)] +
             [identity(2) for n in range(self.num_qubits)])
-        return psi_proj.dag() * U * psi_proj
+        result = psi_proj.dag() * U * psi_proj
+        # In qutip 5 multiplication of matrices
+        # with dims [[1, 2], [2, 2]] and [[2, 2], [1, 2]]
+        # will give a result of
+        # dims [[1, 2], [1, 2]] instead of [[2], [2]].
+        if result.dims[0][0] == 1:
+            result = result.ptrace(list(range(len(self.dims)))[1:])
+        return result
```

Comment:
@jakelishman A quick question if you get time, is this the expected behaviour. The two examples below seem inconsistent on how dimensions are handled.
```python
>>> basis(2,0).dag() * basis(2,0)
(1+0j)

>>> tensor(basis(2,0).dag(), qeye(2)) * tensor(basis(2,0),qeye(2)) 
Quantum object: dims=[[1, 2], [1, 2]], shape=(2, 2), type='oper', isherm=True
Qobj data =
[[1. 0.]
 [0. 1.]]
```

Reply:
This was intended. I think we also added a method called `Qobj.contract_subspaces` or something as well to handle the second behaviour if you wanted it?  (Not certain it actually got put in.)  At any rate, you certainly don't need to do `ptrace`; you can just modify the dimensions.

The point was that automatic Hilbert-space contraction in most cases caused more bugs down the line when reusing the objects, or when dealing with explicit Hilbert spaces of dimension 1 (that may appear programmatically - they're pretty useless in practice).  The contraction to a scalar depends on the operands being of `Qobj` type `bra` and `ket`.  Your example has type `oper`, so you get an `oper` back, and this doesn't contract any Hilbert spaces, because it couldn't tell if the first Hilbert space was `oper * oper` or `bra * ket` or even the forbidden `ket * ket`.

Reply:
> you either get a scalar back if the objects were exactly bra and ket, or you get a Qobj with the same rule for dimensions that everything always uses ([left.dims[0], right.dims[1]]). We had too many problems with attempts to guess when a dimension should be contracted that we just made it exactly consistent in all cases.

Ok, this rule also makes sense. Thanks!